# Analysis of English Premier League

### Content

 1. Introduction: English Premier League
 2. Data description
 3. Research questions
 4. Data preparation: cleaning and shaping

## 1. Introduction: English Premier League

The Premier League, often referred to as the English Premier League or the EPL outside England, is the top level of the English football league system. Contested by 20 clubs, it operates on a system of promotion and relegation with the English Football League (EFL). Seasons run from August to May with each team playing 38 matches (playing all 19 other teams both home and away). Most games are played on Saturday and Sunday afternoons.

The competition was founded as the FA Premier League on 20 February 1992 following the decision of clubs in the Football League First Division to break away from the Football League, founded in 1888, and take advantage of a lucrative television rights deal.

The Premier League is the most-watched sports league in the world, broadcast in 212 territories to 643 million homes and a potential TV audience of 4.7 billion people. For the 2018–19 season average Premier League match attendance was at 38,181,second to the Bundesliga's 43,500, while aggregated attendance across all matches is the highest of any league at 14,508,981. Most stadium occupancies are near capacity. The Premier League ranks second in the UEFA coefficients of leagues based on performances in European competitions over the past five seasons as of 2019, only behind Spain's La Liga.

Source(https://en.wikipedia.org/wiki/Premier_League)

## 2. Data description 

According to the intro information about English Premier League (EPL), it is clear that EPL is the one of the most followed football league in the world. There are popular clubs like Manchester United, Manchester City, Liverpool and etc. It is interesting to discover and compare different features of matches.

Unfortunately, the season 2019-2020 was not passed as usual due to the coronavirus pandemic. Therefore, I decided to take last two seasons to see complete picture. So, analysis will be based on data of 2018-2019 and 2019-2020 season. Below the data that will be used in analysis:

* **Date =** Match Date (dd/mm/yy)
* **HomeTeam =** Home Team
* **AwayTeam =** Away Team
* **FTHG =** Full Time Home Team Goals
* **FTAG =** Full Time Away Team Goals
* **FTR =** Full Time Result (H=Home Team Win, D=Draw, A=Away Team Win)
* **Referee =** Name of the person who judges the match
* **HS =** Home Team Shots
* **AS =** Away Team Shots
* **HST =** Home Team Shots on Target
* **AST =** Away Team Shots on Target
* **HY =** Home Team Yellow Cards
* **AY =** Away Team Yellow Cards
* **HR =** Home Team Red Cards
* **AR =** Away Team Red Cards
* **B365H =** Bet365 home win odds
* **B365D =** Bet365 draw odds
* **B365A =** Bet365 away win odds

## 3. Research questions

In this project, mainly I will focus on several parts of data analysis and visualization:
1. Analyze performance of clubs over seasons
2. Analyze red and yellow cards ratio by referee
3. Analyze the number of goals per match by teams and total goals by match outcome
4. Analyze the shots needed for teams to score a goal
5. Analyze the mean betting odds by match outcome

## 4. Data preparation: cleaning and shaping

At this part, I have almost ready dataset but I should make some operations with them:
 * drop columns that will not be used
 * concatenate two data frames
 * check for the inconsistencies
 * add columns

In [2]:
#import all libraries
import pandas as pd
import numpy as np
import seaborn as sns 
import matplotlib.pyplot as plt

In [6]:
#import datasets
df1 = pd.read_csv("england-premier-league-2018-to-2019.csv")
df2 = pd.read_csv("england-premier-league-2019-to-2020.csv")

#drop nonessential columns
drop = (['Div','HF', 'AF', 'HC', 'AC', 'BWH', 'BWD',
       'BWA', 'IWH', 'IWD', 'IWA', 'PSH', 'PSD', 'PSA', 'WHH', 'WHD', 'WHA',
       'VCH', 'VCD', 'VCA', 'Bb1X2', 'BbMxH', 'BbAvH', 'BbMxD', 'BbAvD',
       'BbMxA', 'BbAvA', 'BbOU', 'BbMx>2.5', 'BbAv>2.5', 'BbMx<2.5',
       'BbAv<2.5', 'BbAH', 'BbAHh', 'BbMxAHH', 'BbAvAHH', 'BbMxAHA', 'BbAvAHA',
       'PSCH', 'PSCD', 'PSCA','HTHG', 'HTAG', 'HTR'])
df1.drop(drop, axis = 'columns', inplace = True)

drop = (['Div', 'Time','HF', 'AF', 'HC', 'AC', 'BWH', 'BWD',
       'BWA', 'IWH', 'IWD', 'IWA', 'PSH', 'PSD', 'PSA', 'WHH', 'WHD', 'WHA',
       'VCH', 'VCD', 'VCA', 'MaxH', 'MaxD', 'MaxA', 'AvgH', 'AvgD',
       'AvgA', 'B365>2.5', 'B365<2.5', 'P>2.5', 'P<2.5', 'Max>2.5', 'Max<2.5',
       'Avg>2.5', 'Avg<2.5', 'AHh', 'B365AHH', 'B365AHA', 'PAHH', 'PAHA',
       'MaxAHH', 'MaxAHA', 'AvgAHH', 'AvgAHA', 'B365CH', 'B365CD', 'B365CA',
       'BWCH', 'BWCD', 'BWCA', 'IWCH', 'IWCD', 'IWCA', 'PSCH', 'PSCD', 'PSCA',
       'WHCH', 'WHCD', 'WHCA', 'VCCH', 'VCCD', 'VCCA', 'MaxCH', 'MaxCD',
       'MaxCA', 'AvgCH', 'AvgCD', 'AvgCA', 'B365C>2.5', 'B365C<2.5', 'PC>2.5',
       'PC<2.5', 'MaxC>2.5', 'MaxC<2.5', 'AvgC>2.5', 'AvgC<2.5', 'AHCh',
       'B365CAHH', 'B365CAHA', 'PCAHH', 'PCAHA', 'MaxCAHH', 'MaxCAHA',
       'AvgCAHH', 'AvgCAHA', 'HTHG', 'HTAG', 'HTR'])
df2.drop(drop, axis = 'columns', inplace = True)

I dropped from two data frames columns that will not be used in further investigation.

Below I concatenated two data frames in order to analyze matches of last 2 seasons.

In [7]:
#Concatenate data frames
frames = [df1, df2]
df = pd.concat(frames, axis = 0, join = 'outer', ignore_index = True)
df.tail(3)

Date  HomeTeam        AwayTeam  FTHG  FTAG FTR     Referee  HS  AS  \
585  01/01/2020   Norwich  Crystal Palace     1     1   D      J Moss  15  12   
586  01/01/2020  West Ham     Bournemouth     4     0   H     G Scott  14   3   
587  01/01/2020   Arsenal      Man United     2     0   H  C Kavanagh  10  10   

     HST  AST  HY  AY  HR  AR  B365H  B365D  B365A  
585    4    3   5   0   0   0   2.50   3.40   2.75  
586    7    2   1   2   0   0   1.90   3.75   3.80  
587    4    4   2   0   0   0   2.55   3.60   2.62

In [8]:
#Checking for missing values
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 588 entries, 0 to 587
Data columns (total 18 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Date      588 non-null    object 
 1   HomeTeam  588 non-null    object 
 2   AwayTeam  588 non-null    object 
 3   FTHG      588 non-null    int64  
 4   FTAG      588 non-null    int64  
 5   FTR       588 non-null    object 
 6   Referee   588 non-null    object 
 7   HS        588 non-null    int64  
 8   AS        588 non-null    int64  
 9   HST       588 non-null    int64  
 10  AST       588 non-null    int64  
 11  HY        588 non-null    int64  
 12  AY        588 non-null    int64  
 13  HR        588 non-null    int64  
 14  AR        588 non-null    int64  
 15  B365H     588 non-null    float64
 16  B365D     588 non-null    float64
 17  B365A     588 non-null    float64
dtypes: float64(3), int64(10), object(5)
memory usage: 82.8+ KB
None


* object(5): date, Home team, Away team, FTR, Referee
* float(3): B365H, B365A, B365D
* int(10): FTHG, FTAG, HS, AS, HST, AST, HY, AY, HR, AR

In [9]:
df.isnull().any()

Date        False
HomeTeam    False
AwayTeam    False
FTHG        False
FTAG        False
FTR         False
Referee     False
HS          False
AS          False
HST         False
AST         False
HY          False
AY          False
HR          False
AR          False
B365H       False
B365D       False
B365A       False
dtype: bool

As it can be seen from the results above, data frame consists of 588 observations and 18 columns. Also there is no null values in any of this columns and type of the columns are proper.

In [28]:
#checking for duplicates 
print(df.duplicated())
print(df.duplicated().sum())

0      False
1      False
2      False
3      False
4      False
       ...  
583    False
584    False
585    False
586    False
587    False
Length: 588, dtype: bool
0


From the results above, I can conclude that there is no any inconsistencies related with data frame.

In [14]:
#add columns since it will be useful later

#total yellow card in match
df['TY'] = df['HY'] + df['AY']

#total red card in match
df['TR'] = df['HR'] + df['AR']

#total goals scored in match
df['TG'] = df['FTHG'] + df['FTAG']

#check
df.head(2)

Date     HomeTeam   AwayTeam  FTHG  FTAG FTR     Referee  HS  AS  \
0  10/08/2018   Man United  Leicester     2     1   H  A Marriner   8  13   
1  11/08/2018  Bournemouth    Cardiff     2     0   H    K Friend  12  10   

   HST  ...  HY  AY  HR  AR  B365H  B365D  B365A  TY  TR  TG  
0    6  ...   2   1   0   0   1.57    3.9    7.5   3   0   3  
1    4  ...   1   1   0   0   1.90    3.6    4.5   2   0   2  

[2 rows x 21 columns]

I added some columns which may give me more insights in data visualization part.